text/0000-newtypes.md

```diff
@@ -0,0 +1,263 @@
+
+// We want to do generic manipulations
+fn calc_distance<T: Sub>(start: T, end: T) -> T {
+    end - start
+}
+
+// Initialize the same way as the underlying types
+let (start_inch, end_inch): (Inch, Inch) = (10, 18);
+let (start_cm, end_cm): (Cm, Cm) = (2, 5);
+
+// Here `calc_distance` operates on the types `Inch` and `Cm`,
+// where previously we had to cast to and from `usize`.
+let inch_dist = calc_distance(start_inch, end_inch);
+let cm_dist = calc_distance(start_cm, end_cm);
+
+println!("dist: {} and {}", inch_dist, cm_dist);
+
+// Disallowed at compile time
+let not_allowed = calc_distance(start_inch, end_cm);
+```
+
+It would also allow generics:
+
+```rust
+struct A<N, M> { n: N, m: M }
+type B<T> is new A<usize, T>;
+
+let b = B { n: 2u, m: "this is a T" };
+```
+
+It would not be possible to use the newtype in place of the parent type,
+we would need to resort to traits.
+
+```rust
+fn bad(x: usize) { ... }
+fn good<T: Sub>(x: T) { ... }
+
+type Foo is new usize;
+let a: Foo = 2;
+bad(a); // Not allowed
+good(a); // Ok, Foo implements Sub
+```
+
+## Derived traits
+
+In the derived trait implementations the basetype will be replaced by the newtype.
+
+So for example as `usize` implements `Add<usize>`, `type Inch is new usize`
+would implement `Add<Inch>`.
+
+## Scoping
+
+Newtypes would follow the natural scoping rules:
+
+```rust
+type Inch is new usize; // Not accessible from outside the module
+pub type Cm is new usize; // Accessible
+
+use module::Inch; // Import into scope
+pub use module::Inch; // Re-export
+```
+
+### Reexporting private types
+
+Newtypes are allowed to be newtypes over private types:
+
+```rust
+mod foo {
+    struct Foo;
+    pub type Bar is new Foo;
+}
+let x: foo::Bar = ...; // OK
+let x: foo::Foo = ...; // Not OK
+```
+
+## Casting
+
+Newtypes can explicitly be converted to their base types, and vice versa.
+Implicit conversions are not allowed.
+This is achieved via the `Into` trait, since newtypes automatically implement
+`From<BaseType>` and `From<NewType> for BaseType`. In order to not expose new
+`as` casts, the automatically generated implementation simply contains a
+`transmute`.
+
+```rust
+type Inch is new usize;
+
+fn frobnicate(x: usize) -> usize { x * 2 + 14 - 3 * x * x }
+
+let x: Inch = 2;
+println!("{}", frobnicate(x.into()));
+
+let a: usize = 2;
+let i: Inch = a; // Compile error, implicit conversion not allowed
+let i = Inch::from(a); // Ok
+let i: Inch = a.into(); // Ok
+let b = usize::from(i); // Ok
+```
+
+## Grammar
+
+The grammar rules will be the same as for `type`, but there are two new
+contextual keywords `is` and `new`. The reason for using `is new` instead of
+another sigil is that `type X = Y;` would be very hard to distinguish from any
+alternative like `type X <- Y;` or just `type X is Y;`.
+
+## Implementation
+
+The compiler would treat newtypes as a thin wrapper around the original type.
+This means that just declaring a newtype does *not* generate any code, because
```

Comment:
This is inconsistent with:
> This is achieved via the `Into` trait, since newtypes automatically implement `From<BaseType>` and `From<NewType> for BaseType`.

Reply:
Ha, good catch 😆